Suggestion for CLI: Refactor variadic command into Click command group, set up tests

alextremblay opened the pull request:
Hi Russell, Here's my proposal for refactoring the CLI.

Currently, the CLI is one click command. I say it's a variadic command, because it's a command that does completely different things depending on its arguments (ie, what 'action' argument you call it with)
What I'm suggesting to do here is to break it up into a set of different commands in a command group. This gives us a number of benefits, some of which are outlined below:
 - Simplifies the CLI code. Each action type has its own function and can focus one what it does best, can have its own set or combination of input arguments, and can do its own validation of those arguments.
 - Enables future support for shell completion (ie. bash tab completion of zsh completion)
 - Makes it easier to separate global command-line options (like `--verbose` and `--silent` with command-specific options and arguments)
 - Makes it easier to test individual commands in isolation

Another change included in this proposal is that I'm set up preliminary test scenarios to test the CLI and define expected behaviour. These tests are written to be run with `pytest` (`pip install pytest`), which is a common and popular python testing framework.

Right now these tests don't make any kind of assertions about expected behaviour, so they're not what I would call 'complete', but they do print out the commands they run and the results they get in different scenarios, and so they're very useful in highlighting how the CLI currently behaves.

alextremblay committed:
Refactor CLI, setup preliminary tests

Refactor the CLI frontend into a Click command group of related commands
Set up preliminary tests for the cli. No value assertions yet, but this adds tests for common scenarios in preparation for writing test assertions

alextremblay committed:
merge upstream changes

quistian commented:
Good. Let me look at it tonight and review your ideas. Thanks again!

Russell P. Sutherland           Email: russell . sutherland @ utoronto dawt ca
Network Engineer, I+TS       Voice: +1.416.978.0470
4 Bancroft Ave., Rm. 102      Cell: +1.416.803.0080
University of Toronto            Fax:   +1.416.978.6620
Toronto, ON  M5S 1C1
________________________________
From: Alex Tremblay <notifications@github.com>
Sent: Monday, April 29, 2019 15:55
To: quistian/bamcli
Cc: Subscribed
Subject: [quistian/bamcli] Suggestion for CLI: Refactor variadic command into Click command group, set up tests (#2)


Hi Russell, Here's my proposal for refactoring the CLI.

Currently, the CLI is one click command. I say it's a variadic command, because it's a command that does completely different things depending on its arguments (ie, what 'action' argument you call it with)
What I'm suggesting to do here is to break it up into a set of different commands in a command group. This gives us a number of benefits, some of which are outlined below:

  *   Simplifies the CLI code. Each action type has its own function and can focus one what it does best, can have its own set or combination of input arguments, and can do its own validation of those arguments.
  *   Enables future support for shell completion (ie. bash tab completion of zsh completion)
  *   Makes it easier to separate global command-line options (like --verbose and --silent with command-specific options and arguments)
  *   Makes it easier to test individual commands in isolation

Another change included in this proposal is that I'm set up preliminary test scenarios to test the CLI and define expected behaviour. These tests are written to be run with pytest (pip install pytest), which is a common and popular python testing framework.

Right now these tests don't make any kind of assertions about expected behaviour, so they're not what I would call 'complete', but they do print out the commands they run and the results they get in different scenarios, and so they're very useful in highlighting how the CLI currently behaves.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/quistian/bamcli/pull/2

Commit Summary

  *   Refactor CLI, setup preliminary tests
  *   merge upstream changes

File Changes

  *   M setup.cfg<https://github.com/quistian/bamcli/pull/2/files#diff-0> (4)
  *   M src/bluecat_am/cli.py<https://github.com/quistian/bamcli/pull/2/files#diff-1> (259)
  *   A tests/test_cli.py<https://github.com/quistian/bamcli/pull/2/files#diff-2> (95)

Patch Links:

  *   https://github.com/quistian/bamcli/pull/2.patch
  *   https://github.com/quistian/bamcli/pull/2.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/quistian/bamcli/pull/2>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABMHKHUKY64LWCZA64UFWMDPS5HC5ANCNFSM4HJGIOQQ>.

alextremblay commented:
An example of what the tests produce when I run `py.test -s` from project root:

```text
========================================================================================================= test session starts =========================================================================================================
platform darwin -- Python 3.7.2, pytest-4.4.1, py-1.8.0, pluggy-0.9.0 -- /Users/alex/.pyenv/versions/bluecat_am/bin/python
cachedir: .pytest_cache
rootdir: /Users/alex/EIS-SI/projects/bamcli, inifile: setup.cfg, testpaths: tests
plugins: pycharm-0.5.0
collected 5 items

tests/test_cli.py::test_cli_command_group
Ran 'bamcli --help'
Got:
Usage: run [OPTIONS] COMMAND [ARGS]...

  Command line interface to BAM DNS System

  E.g.  $bamcli add bozo.uoft.ca A 3600 10.10.10.1 [TTL]

Options:
  -s, --silent   Minimize the output from commands. Silence is golden
  -v, --verbose  Show what is going on for debugging purposes
  --help         Show this message and exit.

Commands:
  add      Add a new Resource Record to the BlueCat DNS system
  delete   Delete a Resource Record in the BlueCat DNS System
  find     Find a Resource Record in the BlueCat DNS System
  list     Alias for the 'view' command
  remove   Alias for the 'delete' command
  replace  Alias for the 'update' command
  update   Update a Resource Record in the BlueCat DNS System
  view     View a Resource Record in the BlueCat DNS System

PASSED
tests/test_cli.py::test_multiple_A_records
Ran 'bamcli add alex.utoronto.ca A 10.128.0.10'
Got:
added new A record:
alex.utoronto.ca     IN       A      10.128.0.10

Ran 'bamcli add alex.utoronto.ca A 10.128.0.20'
Got:
Added Host Record:
added new A record:
alex.utoronto.ca     IN       A      10.128.0.20
alex.utoronto.ca     IN       A      10.128.0.10

Ran 'bamcli add alex.utoronto.ca A 10.128.0.30'
Got:
Added Host Record:
added new A record:
alex.utoronto.ca     IN       A      10.128.0.30
alex.utoronto.ca     IN       A      10.128.0.10
alex.utoronto.ca     IN       A      10.128.0.20

Ran 'bamcli view alex.utoronto.ca A'
Got:
{'id': 0, 'name': None, 'type': None, 'properties': None}
alex.utoronto.ca     IN       A      10.128.0.30
alex.utoronto.ca     IN       A      10.128.0.10
alex.utoronto.ca     IN       A      10.128.0.20

Ran 'bamcli find alex.utoronto.ca A'
Got:
[2525103]

Ran 'bamcli delete alex.utoronto.ca A 10.128.0.30'
Got:
This RR exists and will be deleted

Ran 'bamcli delete alex.utoronto.ca A 10.128.0.20'
Got:
This RR exists and will be deleted

Ran 'bamcli delete alex.utoronto.ca A 10.128.0.10'
Got:
This RR exists and will be deleted

PASSED
tests/test_cli.py::test_add_ip_outside_valid_network
Ran 'bamcli add alex.utoronto.ca A 10.20.30.40'
Got:
Cannot add: .alex.utoronto.ca of type A Error: ["IP Address doesn't belong to a Network.  Could not locate the network."]

PASSED
tests/test_cli.py::test_add_fqdn_outside_range
Ran 'bamcli add something.some.thing A 10.128.0.10'
Got:
Top level domain name must be one of: ('ca', 'com', 'edu', 'org')
Top level domain name must be one of: ('ca', 'com', 'edu', 'org')
Cannot add: something.some.thing of type A Error: ['Invalid destination Zone.']

PASSED
tests/test_cli.py::test_delete_non_existant_FQDN
Ran 'bamcli delete blah.alex.utoronto.ca A 10.128.0.10'
Got:
No RR exists matching blah.alex.utoronto.ca A 10.128.0.10

PASSED

====================================================================================================== 5 passed in 5.55 seconds =======================================================================================================

```

quistian commented:
All good. Very good. My next step was to pass some context, because as you mention different commands have different options. I also had been looking at the testing framework: pytest. You've got me started with some examples. All systems go.
